SparkPost/2web2ui
# SA-1885 - (UI) Address comparator filters w/ same key are dropped by adding form validation

[SA-1885](https://sparkpost.atlassian.net/browse/SA-1885)

### What Changed

- Renamed spec files within `cypress/tests/integration/signals-analytics/analytics-report` to improve consistency
- Refactored `useFiltersForm` to use a `status` enum to help manage state
- Removes `filterByCollectionValues` helper that is no longer used in the app (I just happened to run in to this while looking at our existing array helpers)
- Adds a new error banner that renders at the top of each group that has a duplicate violation (when the user adds filters that have duplicate "Type" and "Compare By" values within the same grouping, i.e. "Campaign" "contains" and then another filter for "Campaign" "contains")


### How To Test

1. Go to `/signals/analytics` and click on "Add Filters"
1. Select "Type" of "Campaign" and in the `<select />` that appears next to it, select "contains"
1. Enter any value within the "Campaign" field and hit enter
1. Click the "AND +" button _within that same grouping_
1. _Within that same grouping_ select the exact same values for "Type" and "Compare By" ("Campaign" and "contains")
1. Try to submit the form
1. Verify the error banner renders at the top of the relevant filter grouping
1. Verify that focus is moved to the error (this can be verified via the console via `document.activeElement`)
1. Afterwards, verify the error clears when:
    -  Swapping the filter "Type" for either of the offending filters
    -  Swapping the filter "Compare By" for either of the offending filters
    -  Clearing the form via the "Clear Filters" button
1. Add an additional filter grouping (By clicking "AND +" outside of the Panel but within the drawer) and introduce another duplication error
1. Submit the form and verify focus moves to the _first error that is still present_
1. Verify that this error clears independently of the first and vice versa
1. It would be worth testing more complex scenarios as well - the error will render (but only once!) when the user has duplicates within _any_ grouping in the form - whoever picks up the review, please try and break this thing.

### To Do

- [x] Run design/copy by UX - [relevant Slack thread](https://msys.slack.com/archives/GLLAW6UAX/p1610044824129000)
- [x] Add automated testing
